src/render.rs

```diff
@@ -310,18 +311,106 @@ impl Render for ServerNoticeMessageEventContent {
     }
 }
 
+/// Convert a matrix content URI to HTTP(s), respecting a user's homeserver
+fn mxc_to_http(
+    mxc_url: &str,
+    homeserver: &Url,
+) -> Result<String, Box<dyn std::error::Error>> {
+    let url = url::Url::parse(mxc_url)?;
+
+    if url.scheme() != "mxc" {
+        return Err("URL missing MXC scheme".into());
+    }
+
+    if url.path().is_empty() {
+        return Err("URL missing path".into());
+    }
+
+    Ok(format!(
+        "{homeserver}_matrix/media/r0/download/{server_name}{media_id}",
```

Comment:
I think using [join](https://docs.rs/url/2.2.1/url/struct.Url.html#method.join) for this is a better idea.